File: TI1.4.RESTFUL_DOTNET_SINBDD_GR10/02. CLIMOV/ClimovDotnetRestful/app/src/main/java/ec/edu/monster/climov_dotnet_soap/api/SoapClient.kt
package ec.edu.monster.climov_dotnet_soap.api

import ec.edu.monster.climov_dotnet_soap.models.ServerConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class SoapClient {
    companion object {
        private const val CONTENT_TYPE = "text/xml; charset=utf-8"
        private const val NAMESPACE = "http://tempuri.org/"
        private const val SERVICE_INTERFACE = "IConversionService"
    }

    suspend fun callWithFailover(methodName: String, paramValue: Double, callback: SoapCallback) =
        withContext(Dispatchers.IO) {
            var lastErr: String? = null
            for (ip in ServerConfig.SERVER_IPS) {
                try {
                    val resp = call(ServerConfig.endpoint(ip), buildSoapEnvelope(methodName, paramValue))
                    withContext(Dispatchers.Main) { callback.onSuccess(resp) }
                    return@withContext
                } catch (e: Exception) {
                    lastErr = e.message ?: "Error desconocido"
                }
            }
            withContext(Dispatchers.Main) {
                callback.onError("No se pudo conectar a ninguna IP. Detalle: $lastErr")
            }
        }

    private fun buildSoapEnvelope(methodName: String, value: Double): String {
        val paramName = when (methodName) {
            "CentimetrosAMetros" -> "centimetros"
            "MetrosACentimetros" -> "metros"
            "MetrosAKilometros" -> "metros"
            "ToneladaALibra" -> "tonelada"
            "KilogramoALibra" -> "kilogramo"
            "GramoAKilogramo" -> "gramos"
            "CelsiusAFahrenheit" -> "celsius"
            "FahrenheitACelsius" -> "fahrenheit"
            "CelsiusAKelvin" -> "celsius"
            else -> "value"
        }
        return """
            <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
                <s:Body>
                    <$methodName xmlns="$NAMESPACE">
                        <$paramName>$value</$paramName>
                    </$methodName>
                </s:Body>
            </s:Envelope>
        """.trimIndent()
    }

    private fun call(endpointUrl: String, body: String): String {
        val url = URL(endpointUrl)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doInput = true
            doOutput = true
            connectTimeout = ServerConfig.CONNECTION_TIMEOUT
            readTimeout = ServerConfig.READ_TIMEOUT
            setRequestProperty("Content-Type", CONTENT_TYPE)
            setRequestProperty("SOAPAction", "\"$NAMESPACE$SERVICE_INTERFACE/${methodNameFromBody(body)}\"")
        }

        OutputStreamWriter(conn.outputStream).use { it.write(body); it.flush() }

        if (conn.responseCode != HttpURLConnection.HTTP_OK) {
            conn.disconnect()
            throw RuntimeException("HTTP ${conn.responseCode}")
        }

        val sb = StringBuilder()
        BufferedReader(InputStreamReader(conn.inputStream)).use { r ->
            var line: String?
            while (r.readLine().also { line = it } != null) sb.append(line)
        }
        conn.disconnect()
        return sb.toString()
    }

    private fun methodNameFromBody(body: String): String {
        val start = body.indexOf('<')
        val end = body.indexOf(' ', start)
        return if (end > start) body.substring(start + 1, end) else "Unknown"
    }

    fun extractResult(xml: String, methodName: String): String? {
        val startTag = "<${methodName}Result>"
        val start = xml.indexOf(startTag)
        if (start == -1) return null
        val end = xml.indexOf("</", start + startTag.length)
        if (end == -1) return null
        return xml.substring(start + startTag.length, end).trim()
    }
}

interface SoapCallback {
    fun onSuccess(xmlResponse: String)
    fun onError(error: String)
}

File: PROYECTOS_SIN_BDD/TI1.2 SOAP_DOTNET_SIN_BD_GR10/02. CLIWEB/soap-web/src/services/soapClient.js
import axios from 'axios';

const SOAP_URL = '/ConversionService.svc';
const NAMESPACE = 'http://tempuri.org/';
const SERVICE_INTERFACE = 'IConversionService'; // ← ¡Importante!

/**
 * Mapea el nombre del método al nombre del parámetro
 */
const getParamName = (methodName) => {
  const paramMap = {
    CentimetrosAMetros: 'centimetros',
    MetrosACentimetros: 'metros',
    MetrosAKilometros: 'metros',
    ToneladaALibra: 'tonelada',
    KilogramoALibra: 'kilogramo',
    GramoAKilogramo: 'gramos',
    CelsiusAFahrenheit: 'celsius',
    FahrenheitACelsius: 'fahrenheit',
    CelsiusAKelvin: 'celsius'
  };
  return paramMap[methodName] || 'value';
};

export const callSoapMethod = async (methodName, value) => {
  const paramName = getParamName(methodName);
  // SOAPAction CORRECTO para WCF: "http://tempuri.org/IConversionService/CentimetrosAMetros"
  const soapAction = `"${NAMESPACE}${SERVICE_INTERFACE}/${methodName}"`;

  const soapBody = `
    <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
      <s:Body>
        <${methodName} xmlns="${NAMESPACE}">
          <${paramName}>${value}</${paramName}>
        </${methodName}>
      </s:Body>
    </s:Envelope>
  `.trim();

  try {
    const response = await axios.post(SOAP_URL, soapBody, {
      headers: {
        'Content-Type': 'text/xml; charset=utf-8',
        'SOAPAction': soapAction // ← ¡Esta es la clave!
      }
    });

    // Extraer el resultado: <CentimetrosAMetrosResponse><CentimetrosAMetrosResult>2.22</CentimetrosAMetrosResult></CentimetrosAMetrosResponse>
    const xml = response.data;
    const resultTag = `<${methodName}Result>`;
    const start = xml.indexOf(resultTag) + resultTag.length;
    const end = xml.indexOf('</', start);
    
    if (start > -1 && end > start) {
      const result = xml.substring(start, end);
      return parseFloat(result);
    }

    throw new Error('No se encontró el resultado en la respuesta SOAP');
  } catch (error) {
    console.error('SOAP Error:', error);
    if (error.response) {
      console.error('Respuesta del servidor:', error.response.data);
    }
    throw new Error('Error al comunicarse con el servicio SOAP. Verifique que el servidor esté activo.');
  }
};